model/textparse/protobufparse.go

```diff
@@ -422,6 +423,16 @@ func (p *ProtobufParser) Next() (Entry, error) {
 		default:
 			return EntryInvalid, fmt.Errorf("unknown metric type for metric %q: %s", name, p.mf.GetType())
 		}
+		unit := p.mf.GetUnit()
+		if len(unit) > 0 {
+			sfx := fmt.Sprintf("_%s", unit)
```

Comment:
I haven't benchmarked it (maybe you have?), but I would assume that this line is relatively costly because it creates a whole new string. The parser _is_ a fairly hot code path as everything Prometheus scrapes goes through it.

The OpenMetrics parses uses an expression like `!strings.HasSuffix(m, u) || len(m) < len(u)+1 || p.l.b[p.offsets[1]-len(u)-1] != '_'`, which is of course rather cryptic, but it avoids the string creation. Personall, I would use more descriptive variable names (as you did here), and then you could pretty much do the same thing here, like
```Go
if !strings.HasSuffix(name, unit) || len(name) < len(unit)+1 || name[len(name)-len(unit)-1] != '_' {
   return EntryInvalid ...
```

For the counter version, it gets a bit more complex, using `name[:len(name)-6]` instead of `name` itself to cut out the `_total` suffix.